Tests/UpstreamIntegrationTests/util/AnyCodableUtils.swift

```diff
@@ -0,0 +1,733 @@
+//
+// Copyright 2023 Adobe. All rights reserved.
+// This file is licensed to you under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License. You may obtain a copy
+// of the License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under
+// the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+// OF ANY KIND, either express or implied. See the License for the specific language
+// governing permissions and limitations under the License.
+//
+
+import AEPServices
+import Foundation
+import XCTest
+
+enum AssertMode {
+    case exactMatch
+    case typeMatch
+}
+
+extension XCTestCase {
+    // MARK: - AnyCodable exact equivalence test assertion methods
+
+    /// Performs testing assertions between two `AnyCodable` instances, using a similar logic path as the `AnyCodable ==` implementation.
+    /// Traces the key path (both dictionary keys and array indices) and provides the trace on assertion failure, for easier debugging.
+    /// Automatically performs any required conversions of underlying `Any?` types into `AnyCodable` format.
+    ///
+    /// Main entrypoint for `AnyCodable` testing assertions.
+    @discardableResult
+    func assertEqual(expected: AnyCodable?, actual: AnyCodable?, keyPath: [Any] = [], file: StaticString = #file, line: UInt = #line, shouldAssert: Bool = true) -> Bool {
```

Comment:
I don't see tests for assertEqual with keyPath, will we use this option or will assertContains cover all the cases? If the latter, then maybe we can just make one API public assertEqual(expected, actual, file, line) and keep all the other helpers used for internal computation and recursion private